From 30c70196293e8acea78eb55d3ce1ea40614804dc Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Wed, 04 Nov 2020 14:13:22 +0200
Subject: [PATCH] [SW-12755] gaudi: security: generalize pcie rtr disable

As a preparation for securing dynamically the versions and FIT images,
generalize the PCIe RTR disabling function to work on any provided RR.
This will enable opening the PCIe RTR dynamic RR settings.

Change-Id: Ibc91206c916a6f9711188e21af790a12a69d72d1
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/gaudi/security.c b/gaudi/security.c
index 388f168..9dafd8b 100644
--- a/gaudi/security.c
+++ b/gaudi/security.c
@@ -1265,12 +1265,36 @@
 	preboot_secure_pcie_rr(16, mmSTM_BASE - 1, mmRAZWI_BASE);
 }
 
+static void preboot_security_pci_rtr_disable(int rr)
+{
+	u32 reg = rr * sizeof(u32);
+	u32 hit;
+
+	/* PCIe RTR RR - HBM WP 0x0 - 0x07FFFFFFFF */
+	regmap_writel(0, reg + mmSIF_RTR_CTRL_7_RANGE_PRIV_BASE_LOW_AW_0);
+	regmap_writel(0, reg + mmSIF_RTR_CTRL_7_RANGE_PRIV_BASE_HIGH_AW_0);
+	regmap_writel(0, reg + mmSIF_RTR_CTRL_7_RANGE_PRIV_MASK_LOW_AW_0);
+	regmap_writel(0, reg + mmSIF_RTR_CTRL_7_RANGE_PRIV_MASK_HIGH_AW_0);
+	/* PCIe RTR RR - Disable (hit) RR */
+	hit = regmap_readl(mmSIF_RTR_CTRL_7_RANGE_PRIV_HIT_AW);
+	regmap_writel(hit & ~BIT(rr), mmSIF_RTR_CTRL_7_RANGE_PRIV_HIT_AW);
+
+	/* PCIe RTR RR - HBM RP 0x0 - 0x07FFFFFFFF */
+	regmap_writel(0, reg + mmSIF_RTR_CTRL_7_RANGE_PRIV_BASE_LOW_AR_0);
+	regmap_writel(0, reg + mmSIF_RTR_CTRL_7_RANGE_PRIV_BASE_HIGH_AR_0);
+	regmap_writel(0, reg + mmSIF_RTR_CTRL_7_RANGE_PRIV_MASK_LOW_AR_0);
+	regmap_writel(0, reg + mmSIF_RTR_CTRL_7_RANGE_PRIV_MASK_HIGH_AR_0);
+	/* PCIe RTR RR - Disable (hit) RR */
+	hit = regmap_readl(mmSIF_RTR_CTRL_7_RANGE_PRIV_HIT_AR);
+	regmap_writel(hit & ~BIT(rr), mmSIF_RTR_CTRL_7_RANGE_PRIV_HIT_AR);
+}
+
 #define PCIE_WRAP_RRS_NUM	32
 
 static void preboot_security_pci_disable(void)
 {
 	int i;
-	u32 hit, base, prot_base;
+	u32 base, prot_base;
 
 	for (i = 0; i < PCIE_WRAP_RRS_NUM; i++) {
 		base = i * sizeof(u32);
@@ -1284,23 +1308,7 @@
 		regmap_writel(0, prot_base);
 	}
 
-	/* PCIe RTR RR - HBM WP 0x0 - 0x07FFFFFFFF */
-	regmap_writel(0, mmSIF_RTR_CTRL_7_RANGE_PRIV_BASE_LOW_AW_0);
-	regmap_writel(0, mmSIF_RTR_CTRL_7_RANGE_PRIV_BASE_HIGH_AW_0);
-	regmap_writel(0, mmSIF_RTR_CTRL_7_RANGE_PRIV_MASK_LOW_AW_0);
-	regmap_writel(0, mmSIF_RTR_CTRL_7_RANGE_PRIV_MASK_HIGH_AW_0);
-	/* PCIe RTR RR - Disable (hit) RR */
-	hit = regmap_readl(mmSIF_RTR_CTRL_7_RANGE_PRIV_HIT_AW);
-	regmap_writel(hit & ~BIT(0), mmSIF_RTR_CTRL_7_RANGE_PRIV_HIT_AW);
-
-	/* PCIe RTR RR - HBM RP 0x0 - 0x07FFFFFFFF */
-	regmap_writel(0, mmSIF_RTR_CTRL_7_RANGE_PRIV_BASE_LOW_AR_0);
-	regmap_writel(0, mmSIF_RTR_CTRL_7_RANGE_PRIV_BASE_HIGH_AR_0);
-	regmap_writel(0, mmSIF_RTR_CTRL_7_RANGE_PRIV_MASK_LOW_AR_0);
-	regmap_writel(0, mmSIF_RTR_CTRL_7_RANGE_PRIV_MASK_HIGH_AR_0);
-	/* PCIe RTR RR - Disable (hit) RR */
-	hit = regmap_readl(mmSIF_RTR_CTRL_7_RANGE_PRIV_HIT_AR);
-	regmap_writel(hit & ~BIT(0), mmSIF_RTR_CTRL_7_RANGE_PRIV_HIT_AR);
+	preboot_security_pci_rtr_disable(0);
 }
 
 void preboot_security_init(void)
